plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiRecordCursor.java

```diff
@@ -0,0 +1,163 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.loki;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.trino.spi.block.SqlMap;
+import io.trino.spi.connector.RecordCursor;
+import io.trino.spi.type.MapType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeUtils;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.spi.block.MapValueBuilder.buildMapValue;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_NANOS;
+import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.util.Objects.requireNonNull;
+
+public class LokiRecordCursor
+        implements RecordCursor
+{
+    final List<LokiColumnHandle> columnHandles;
+    final LokiQueryResultIterator resultIterator;
+
+    private boolean advanced;
+
+    public LokiRecordCursor(List<LokiColumnHandle> columnHandles, LokiQueryResultIterator queryResult)
+    {
+        requireNonNull(columnHandles, "columnHandles is null");
+        requireNonNull(queryResult, "queryResult is null");
+
+        this.resultIterator = queryResult;
+        this.columnHandles = columnHandles.stream()
+                .sorted(Comparator.comparing(LokiColumnHandle::ordinalPosition))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        this.advanced = true;
+        return this.resultIterator.advanceNextPosition();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return columnHandles.get(field).type();
+    }
+
+    Object getFieldValue(int field)
+    {
+        checkState(this.advanced, "Cursor has not been advanced yet");
+
+        if (field < 0 || field >= columnHandles.size()) {
+            throw new IllegalArgumentException("Invalid field index: " + field);
+        }
+
+        LokiColumnHandle column = this.columnHandles.get(field);
+        return switch (column.name()) {
+            case "labels" -> getSqlMapFromLabels((MapType) column.type(), this.resultIterator.getLabels());
```

Comment:
The label is a same for a given Iterator type ? So in this case do we need to build them each time for each row ? Can we reuse them ?